server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotsServiceIT.java

```diff
@@ -111,6 +115,72 @@ public void testSnapshotDeletionFailureShouldBeLogged() throws Exception {
+        repository.blockOnDataFiles();
+        try {
+            clusterAdmin().prepareDeleteSnapshot("test-repo", "test-snapshot").setWaitForCompletion(false).execute(listener);
+            // The request will complete as soon as the deletion is scheduled
+            safeGet(listener);
+            // The deletion won't complete until the block is removed
+            assertFalse(snapshotDeletionListener.isDone());
+        } finally {
+            repository.unblock();
+        }
+        safeAwait(snapshotDeletionListener);
+    }
+
+    public void testDeleteSnapshotWhenWaitingForCompletion() throws Exception {
+        createIndexWithRandomDocs("test-index", randomIntBetween(1, 5));
+        createRepository("test-repo", "mock");
+        createSnapshot("test-repo", "test-snapshot", List.of("test-index"));
+        MockRepository repository = getRepositoryOnMaster("test-repo");
+        PlainActionFuture<AcknowledgedResponse> requestCompleteListener = new PlainActionFuture<>();
+        SubscribableListener<Void> snapshotDeletionListener = createSnapshotDeletionListener("test-repo");
+        repository.blockOnDataFiles();
+        try {
+            clusterAdmin().prepareDeleteSnapshot("test-repo", "test-snapshot").setWaitForCompletion(true).execute(requestCompleteListener);
+            // Neither the request nor the deletion will complete until we remove the block
+            assertFalse(requestCompleteListener.isDone());
+            assertFalse(snapshotDeletionListener.isDone());
+        } finally {
+            repository.unblock();
+        }
+        safeGet(requestCompleteListener);
+        safeAwait(snapshotDeletionListener);
+    }
+
+    /**
+     * Create a listener that completes once it has observed a snapshot delete begin and end for a specific repository
+     *
+     * @param repositoryName The repository to monitor for deletions
+     * @return the listener
+     */
+    private SubscribableListener<Void> createSnapshotDeletionListener(String repositoryName) {
+        AtomicBoolean deleteHasStarted = new AtomicBoolean(false);
+        return ClusterServiceUtils.addTemporaryStateListener(
+            internalCluster().getCurrentMasterNodeInstance(ClusterService.class),
+            state -> {
+                SnapshotDeletionsInProgress deletionsInProgress = (SnapshotDeletionsInProgress) state.getCustoms()
+                    .get(SnapshotDeletionsInProgress.TYPE);
+                if (deletionsInProgress == null) {
+                    return false;
+                }
+                if (deleteHasStarted.get() == false) {
+                    deleteHasStarted.set(deletionsInProgress.hasExecutingDeletion(repositoryName));
+                    return false;
+                } else {
+                    return deletionsInProgress.hasExecutingDeletion(repositoryName) == false;
+                }
+            }
+        );
+    }
```

Comment:
This is pretty long-winded just because of the asynchrony if we added the listener after calling delete we'd need to cover the cases of
- SnapshotDeletionsInProgress is null
- The delete hasn't started
- The delete has already completed

So it seemed to make sense to just add it before we call delete and track the whole progression?